File: backend/src/modules/dispatcher/dispatcher.controller.ts
import {
  Controller,
  Get,
  Post,
  Body,
  Patch,
  Param,
  Delete,
  Query,
  ParseIntPipe,
} from '@nestjs/common';
import { ApiTags, ApiOperation, ApiResponse } from '@nestjs/swagger';
import { DispatcherService } from './dispatcher.service';
import { CreateDispatcherDto } from './dto/create-dispatcher.dto';
import { UpdateDispatcherDto } from './dto/update-dispatcher.dto';
import { PaginationDto } from '../../common/dto/pagination.dto';
import { ApiResponseDto } from '../../common/dto/api-response.dto';
import { Dispatcher } from './entities/dispatcher.entity';

@ApiTags('调度器管理')
@Controller('api/dispatchers')
export class DispatcherController {
  constructor(private readonly dispatcherService: DispatcherService) {}

  @Post()
  @ApiOperation({ summary: '创建调度目标' })
  @ApiResponse({ status: 201, description: '创建成功' })
  async create(@Body() createDto: CreateDispatcherDto): Promise<ApiResponseDto<Dispatcher>> {
    const dispatcher = await this.dispatcherService.create(createDto);
    return ApiResponseDto.success(dispatcher, '创建成功');
  }

  @Get()
  @ApiOperation({ summary: '获取调度目标列表' })
  @ApiResponse({ status: 200, description: '获取成功' })
  async findAll(@Query() paginationDto: PaginationDto): Promise<ApiResponseDto> {
    const result = await this.dispatcherService.findAll(paginationDto);
    return ApiResponseDto.success(result);
  }

  @Get('setids')
  @ApiOperation({ summary: '获取所有调度组 ID' })
  @ApiResponse({ status: 200, description: '获取成功' })
  async getSetIds(): Promise<ApiResponseDto<number[]>> {
    const setIds = await this.dispatcherService.getSetIds();
    return ApiResponseDto.success(setIds);
  }

  @Get('stats')
  @ApiOperation({ summary: '获取调度器统计信息' })
  @ApiResponse({ status: 200, description: '获取成功' })
  async getStats(): Promise<ApiResponseDto> {
    const stats = await this.dispatcherService.getStats();
    return ApiResponseDto.success(stats);
  }

  @Get('status')
  @ApiOperation({ summary: '获取调度器实时状态' })
  @ApiResponse({ status: 200, description: '获取成功' })
  async getStatus(): Promise<ApiResponseDto> {
    const status = await this.dispatcherService.getStatus();
    return ApiResponseDto.success(status);
  }

  @Get('setid/:setid')
  @ApiOperation({ summary: '根据调度组获取调度目标' })
  @ApiResponse({ status: 200, description: '获取成功' })
  async findBySetId(
    @Param('setid', ParseIntPipe) setid: number,
  ): Promise<ApiResponseDto<Dispatcher[]>> {
    const dispatchers = await this.dispatcherService.findBySetId(setid);
    return ApiResponseDto.success(dispatchers);
  }

  @Get(':id')
  @ApiOperation({ summary: '获取调度目标详情' })
  @ApiResponse({ status: 200, description: '获取成功' })
  async findOne(@Param('id', ParseIntPipe) id: number): Promise<ApiResponseDto<Dispatcher>> {
    const dispatcher = await this.dispatcherService.findOne(id);
    return ApiResponseDto.success(dispatcher);
  }

  @Patch(':id')
  @ApiOperation({ summary: '更新调度目标' })
  @ApiResponse({ status: 200, description: '更新成功' })
  async update(
    @Param('id', ParseIntPipe) id: number,
    @Body() updateDto: UpdateDispatcherDto,
  ): Promise<ApiResponseDto<Dispatcher>> {
    const dispatcher = await this.dispatcherService.update(id, updateDto);
    return ApiResponseDto.success(dispatcher, '更新成功');
  }

  @Delete(':id')
  @ApiOperation({ summary: '删除调度目标' })
  @ApiResponse({ status: 200, description: '删除成功' })
  async remove(@Param('id', ParseIntPipe) id: number): Promise<ApiResponseDto> {
    await this.dispatcherService.remove(id);
    return ApiResponseDto.success(null, '删除成功');
  }

  @Post('reload')
  @ApiOperation({ summary: '重载调度器配置' })
  @ApiResponse({ status: 200, description: '重载成功' })
  async reload(): Promise<ApiResponseDto> {
    await this.dispatcherService.reload();
    return ApiResponseDto.success(null, '重载成功');
  }
}

File: kam-admin-module-routing/src/main/java/io/kamailio/admin/modules/dispatcher/controller/DispatcherController.java
package io.kamailio.admin.modules.dispatcher.controller;

import io.kamailio.admin.common.dto.ApiResponse;
import io.kamailio.admin.common.dto.PaginatedResult;
import io.kamailio.admin.common.dto.PaginationDto;
import io.kamailio.admin.modules.dispatcher.dto.CreateDispatcherDto;
import io.kamailio.admin.modules.dispatcher.dto.UpdateDispatcherDto;
import io.kamailio.admin.modules.dispatcher.entity.Dispatcher;
import io.kamailio.admin.modules.dispatcher.service.DispatcherService;
import io.swagger.v3.oas.annotations.Operation;
import io.swagger.v3.oas.annotations.tags.Tag;
import jakarta.validation.Valid;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.util.List;
import java.util.Map;

@Tag(name = "调度器管理")
@RestController
@RequestMapping("/api/dispatchers")
public class DispatcherController {

    private final DispatcherService service;

    public DispatcherController(DispatcherService service) {
        this.service = service;
    }

    @PostMapping
    @Operation(summary = "创建调度目标")
    public ResponseEntity<ApiResponse<Dispatcher>> create(@Valid @RequestBody CreateDispatcherDto dto) {
        var entity = service.create(dto);
        return ResponseEntity.ok(ApiResponse.success(entity, "创建成功"));
    }

    @GetMapping
    @Operation(summary = "获取调度目标列表")
    public ResponseEntity<ApiResponse<PaginatedResult<Dispatcher>>> findAll(PaginationDto dto) {
        var result = service.findAll(dto);
        return ResponseEntity.ok(ApiResponse.success(result));
    }

    @GetMapping("/setids")
    @Operation(summary = "获取所有调度组 ID")
    public ResponseEntity<ApiResponse<List<Integer>>> getSetIds() {
        var ids = service.getSetIds();
        return ResponseEntity.ok(ApiResponse.success(ids));
    }

    @GetMapping("/stats")
    @Operation(summary = "获取调度器统计信息")
    public ResponseEntity<ApiResponse<Map<String, Object>>> getStats() {
        var stats = service.getStats();
        return ResponseEntity.ok(ApiResponse.success(stats));
    }

    @GetMapping("/status")
    @Operation(summary = "获取调度器实时状态")
    public ResponseEntity<ApiResponse<Object>> getStatus() {
        var status = service.getStatus();
        return ResponseEntity.ok(ApiResponse.success(status));
    }

    @GetMapping("/setid/{setid}")
    @Operation(summary = "根据调度组获取调度目标")
    public ResponseEntity<ApiResponse<List<Dispatcher>>> findBySetId(@PathVariable Integer setid) {
        var list = service.findBySetId(setid);
        return ResponseEntity.ok(ApiResponse.success(list));
    }

    @GetMapping("/{id}")
    @Operation(summary = "获取调度目标详情")
    public ResponseEntity<ApiResponse<Dispatcher>> findOne(@PathVariable Integer id) {
        var entity = service.findOne(id);
        return ResponseEntity.ok(ApiResponse.success(entity));
    }

    @PatchMapping("/{id}")
    @Operation(summary = "更新调度目标")
    public ResponseEntity<ApiResponse<Dispatcher>> update(
            @PathVariable Integer id,
            @Valid @RequestBody UpdateDispatcherDto dto) {
        var entity = service.update(id, dto);
        return ResponseEntity.ok(ApiResponse.success(entity, "更新成功"));
    }

    @DeleteMapping("/{id}")
    @Operation(summary = "删除调度目标")
    public ResponseEntity<ApiResponse<Void>> remove(@PathVariable Integer id) {
        service.remove(id);
        return ResponseEntity.ok(ApiResponse.success(null, "删除成功"));
    }

    @PostMapping("/reload")
    @Operation(summary = "重载调度器配置")
    public ResponseEntity<ApiResponse<Void>> reload() {
        service.reload();
        return ResponseEntity.ok(ApiResponse.success(null, "重载成功"));
    }

    @PostMapping("/set-state")
    @Operation(summary = "设置调度目标状态 (a=active,i=inactive,p=probing,d=disabled,t=trying)")
    public ResponseEntity<ApiResponse<Void>> setState(@RequestBody SetStateRequest body) {
        service.setState(body.getState(), body.getGroup(), body.getAddress());
        return ResponseEntity.ok(ApiResponse.success(null, "状态设置成功"));
    }

    public static class SetStateRequest {
        private String state;
        private Integer group;
        private String address;

        public String getState() { return state; }
        public void setState(String state) { this.state = state; }
        public Integer getGroup() { return group; }
        public void setGroup(Integer group) { this.group = group; }
        public String getAddress() { return address; }
        public void setAddress(String address) { this.address = address; }
    }
}
